File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode result = null;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        recurseTreeHelper(root, p, q);
        return result;
    }
    private boolean recurseTreeHelper(TreeNode currentNode, TreeNode p, TreeNode q) {
        if(currentNode == null)
            return false;
        int l = recurseTreeHelper(currentNode.left, p, q) ? 1 : 0;
        int r = recurseTreeHelper(currentNode.right, p, q) ? 1 : 0;
        int mid = currentNode == p || currentNode == q ? 1 : 0;
        int total = l + r + mid;
        if (total >= 2)
            result = currentNode;
        return total == 1 ? true : false; 
    }

}

File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def recurse_tree(current):
            if not current:
                return False
            left = recurse_tree(current.left) 
            right = recurse_tree(current.right)
            mid = current == p or current == q
            if mid + left + right >= 2:
                self.result = current
            return left or right or mid
             
        recurse_tree(root)
        return self.result
            
       
        